File: Math/Beam.cpp
#include "Math/Beam.hpp"

namespace Math {
    Beam::Beam(const Math::Ray &ray, const Math::Bivector &originDifferential, const Math::Bivector &directionDifferential)
        : mRay(ray), mOriginDifferential(originDifferential), mDirectionDifferential(directionDifferential)
    {
    }

    Beam::Beam(const BeamProxy &proxy)
    {
        mRay = Ray(proxy.ray);
        mOriginDifferential = Bivector(proxy.originDifferential);
        mDirectionDifferential = Bivector(proxy.directionDifferential);
    }

    const Math::Ray &Beam::ray() const
    {
        return mRay;
    }

    const Math::Bivector &Beam::originDifferential() const
    {
        return mOriginDifferential;
    }

    const Math::Bivector &Beam::directionDifferential() const
    {
        return mDirectionDifferential;
    }

    Math::Bivector Beam::project(float distance, const Math::Normal &normal) const
    {
        const Math::Bivector &dP = mOriginDifferential;
        const Math::Bivector &dD = mDirectionDifferential;
        const Math::Vector &D = mRay.direction();
        const Math::Vector N(normal);
        float t = distance;

        Math::Bivector A = dP + dD * t;
        Math::Bivector dtD = Math::Bivector(D * (A.u() * N), D * (A.v() * N)) / (N * D);

        return A - dtD;
    }

    void Beam::writeProxy(BeamProxy &proxy) const
    {
        mRay.writeProxy(proxy.ray);
        mOriginDifferential.writeProxy(proxy.originDifferential);
        mDirectionDifferential.writeProxy(proxy.directionDifferential);
    }
}

File: src/geo/beam.rs
use crate::geo;

use geo::Ray;
use geo::Bivec3;
use geo::Normal3;

pub struct Beam {
    pub ray : Ray,
    pub origin_diff : Bivec3,
    pub direction_diff : Bivec3
}

impl Beam {
    pub fn new(ray : Ray, origin_diff : Bivec3, direction_diff : Bivec3) -> Beam {
        Beam{ray, origin_diff, direction_diff}
    }

    pub fn project(&self, distance : f32, nrm : Normal3) -> Bivec3 {
        let dp = self.origin_diff;
        let dd = self.direction_diff;
        let d = self.ray.direction;
        let t = distance;

        let a = dp + dd * t;
        let dtd = Bivec3::new(d * (a.u * nrm), d * (a.v * nrm)) / (nrm * d);

        return a - dtd;
    }
}